### PR TITLE
feat(runtime): add promise_yield_create2 host function

### PR DESCRIPTION
**Pull request by graphite**
## Summary
- Add new `promise_yield_create2` host function that allows contracts to provide a custom 32-byte yield ID and timeout when creating yield/resume promises
- Store bidirectional `yield_id ↔ data_id` mappings in the trie for duplicate detection and cleanup on resume/timeout
- Gate behind `YieldCreate2` nightly protocol feature (version 153)

## Motivation
The current `promise_yield_create` returns a runtime-generated `data_id` that contracts must store in state to later call `promise_yield_resume`. For use cases like in-contract mempools (NEP-0616 wallet extensions, Aurora), this storage cost is prohibitive. `promise_yield_create2` lets contracts provide their own deterministic yield ID, avoiding expensive storage.

Context: https://near.zulipchat.com/#narrow/channel/295558-core/topic/yield_create2

## Host function signature
```
promise_yield_create2(
    method_name_len, method_name_ptr,
    arguments_len, arguments_ptr,
    gas, gas_weight,
    yield_id_len, yield_id_ptr,     // user-provided 32-byte yield ID
    yield_timeout_blocks,            // custom timeout (only 200 accepted for now)
    register_id                      // data_id written here
) -> u64  // promise index
```

## Key design decisions
- **data_id generation**: Normal `generate_data_id()`, not derived from yield_id
- **Trie mappings**: `YieldIdToDataId` and `DataIdToYieldId` for duplicate detection and cleanup
- **Duplicate detection**: Errors with `YieldIdAlreadyExists` if yield_id is already pending
- **Timeout**: Only 200 blocks accepted for now; will be expanded later
- **Compatible with existing `promise_yield_resume`** — data_id written to register

## Test plan
- [x] Existing yield/resume integration tests pass (stable)
- [x] New `create2_then_resume` nightly test passes
- [x] New `create2_and_resume_in_one_call` nightly test passes
- [x] Parameter snapshot tests updated and passing
- [x] `cargo check` with nightly features passes

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by mitinarseny** [comment]
Why do you want to return `data_id` in `register_id`? 
> Host function signature
> ```rust
> fn promise_yield_create2(
>     method_name_len, method_name_ptr,
>     arguments_len, arguments_ptr,
>     gas, gas_weight,
>     yield_id_len, yield_id_ptr,     // user-provided 32-byte yield ID
>     yield_timeout_blocks,           // custom timeout (only 200 accepted for now)
>     register_id                     // data_id written here
> ) -> u64  // promise index
> ```

Isn't the motivation to avoid storing anything, so that the contract can resume by the same `yield_id` later via `yield_resume2`? Or, to have `yield_resume_data_id(yield_id)` for `yield_id -> data_id` conversion

```rust
fn promise_yield_resume(
        data_id_len: u64, // <- what to put here?
        data_id_ptr: u64, // <- what to put here?
        payload_len: u64,
        payload_ptr: u64,
    ) -> u32;
```

File: core/parameters/res/runtime_configs/parameters.yaml [hunk 1]
```diff
@@ -281,6 +281,7 @@ reftypes_bulk_memory: false
 deterministic_account_ids: false
 gas_key_host_fns: false
 one_yocto_on_promise: false
+yield_create2_host_fns: false
```
Comment (jancionear) [inline core/parameters/res/runtime_configs/parameters.yaml]: Tbh I would just make this a `ProtocolFeature`, like `ProtocolFeature::YieldResumeImprovements`. Would make things simpler. Although both options work.

Comment (jancionear) [reply core/parameters/res/runtime_configs/parameters.yaml]: Ah nevermind, we can't use a ProtocolFeature because we're adding a new host function, and it needs to be gated behind a runtime parameter

**Comment by graphite** [comment]
@mitinarseny this was an artifact of switching from using `yield_id` as `data_id` directly (can reuse the same resume) to keeping a mapping (new resume is needed). Should be fixed now.

File: runtime/near-vm-runner/src/logic/logic.rs [hunk 2]
```diff
@@ -3435,6 +3435,87 @@ bls12381_p2_decompress_base + bls12381_p2_decompress_element * num_elements`
...
+        method_name_ptr: u64,
+        arguments_len: u64,
+        arguments_ptr: u64,
+        gas: u64,
+        gas_weight: u64,
+        yield_id_len: u64,
+        yield_id_ptr: u64,
+        yield_timeout_blocks: u64,
+        register_id: u64,
+    ) -> Result<u64> {
```
Comment (mitinarseny) [inline runtime/near-vm-runner/src/logic/logic.rs]: What's the purpose of returning `data_id` in a register? Shouldn't it be purely internal to host implementation?

Comment (graphite) [reply runtime/near-vm-runner/src/logic/logic.rs]: Good point, this is a leaky abstraction that only may confuse contract devs, removed.

**Comment by graphite** [comment]
@jancionear will you have time to check it out?

**Comment by jancionear** [comment]
I will take a look tomorrow (Wednesday), unless some urgent bee tasks get in the way

File: core/primitives/src/trie_key.rs [hunk 3]
```diff
@@ -256,6 +264,18 @@ pub enum TrieKey {
         receiver_id: AccountId,
         data_id: CryptoHash,
     } = col::PROMISE_YIELD_STATUS,
+    /// Mapping from user-provided yield ID to runtime-generated data ID.
+    /// Used by `promise_yield_create_with_id` for duplicate detection.
+    YieldIdToDataId {
+        receiver_id: AccountId,
+        yield_id: CryptoHash,
```
Comment (jancionear) [inline core/primitives/src/trie_key.rs]: Would be nice to have a newtype for YieldId (`pub struct YieldId(CryptoHash)`) to make sure we never confuse it with data_id.

File: runtime/near-test-contracts/test-contract-rs/src/lib.rs [hunk 4]
```diff
@@ -198,12 +198,31 @@ extern "C" {
...
+    fn promise_yield_create_with_id(
+        method_name_len: u64,
+        method_name_ptr: u64,
+        arguments_len: u64,
+        arguments_ptr: u64,
+        gas: u64,
+        gas_weight: u64,
+        yield_id_len: u64,
+        yield_id_ptr: u64,
+        yield_timeout_blocks: u64,
```
Comment (jancionear) [inline runtime/near-test-contracts/test-contract-rs/src/lib.rs]: I'm not a big fan of the `yield_timeout_blocks` argument. It locks us into a specific API for a feature that isn't implemented yet. What if in the future we will want to change the timeout logic, maybe allow to specify a timeout in seconds? Also it doesn't work with the older `promise_yield_create` host function.

Maybe we could add some sort of configuration method? Something like:
```rust
// Configure the timeout for the next promise_resume_create[with_id]` call
promise_yield_set_timeout_in_blocks(100);
// Create a PromiseYield receipt with a timeout of 100 blocks
promise_yield_create(...);
```

This would give us more flexibility in the API and make it backwards compatible with the previous functions.

One disadvantage is that it's stateful, which could be error-prone, as a developer I could forget that I modifed some parameters 100 lines before. We could make it so that these parameters apply only to the very next `promise_yield_create `, after that they reset to the default and have to be set again. Then we can wrap that in some nicer syntax in the sdk.

I'd leave the timeouts for later, we can add them as a separate feature once we get around to implementing them.

File: runtime/near-test-contracts/test-contract-rs/src/lib.rs [hunk 5]
```diff
@@ -1290,6 +1309,146 @@ pub unsafe fn call_yield_create_and_resume() {
...
+        payload.as_ptr() as u64,
+        gas_fixed,
+        gas_weight,
+        yield_id.len() as u64,
+        yield_id.as_ptr() as u64,
+        yield_timeout_blocks,
+    );
+
+    // Second call with the same yield_id should abort with YieldIdAlreadyExists
+    promise_yield_create_with_id(
```
Comment (jancionear) [inline runtime/near-test-contracts/test-contract-rs/src/lib.rs]: Do we really want to abort? Maybe we could just return an error? It might be useful for contract developers to know if this yield_id was already used.

File: runtime/runtime/src/function_call.rs [hunk 6]
```diff
@@ -153,13 +153,15 @@ pub(crate) fn action_function_call(
...
+                        .yield_timeout_blocks
+                        .unwrap_or(config.wasm_config.limit_config.yield_timeout_length_in_blocks);
                     enqueue_promise_yield_timeout(
                         state_update,
                         &mut promise_yield_indices,
                         account_id.clone(),
                         receipt.input_data_ids[0],
-                        apply_state.block_height
-                            + config.wasm_config.limit_config.yield_timeout_length_in_blocks,
+                        apply_state.block_height + timeout,
```
Comment (jancionear) [inline runtime/runtime/src/function_call.rs]: This will not work with timeouts that are not exactly 200 blocks. The timeout queue is a FIFO queue, not a priority queue. `resolve_promise_yield_timeouts` keeps taking the first element out of the queue until it encounters an element that doesn't need to be timed out yet, then stops.

For example if the block heights in the queue are: `[100, 104, 50]` and the current height is 101, the code will take out the `100` entry, timeout it, then see that `104 > 101` and stop there. It will not notice the `50` entry.

So doing promise_yield with a timeout at `200` and another one with timeout `100` will not work, because the `200` entry will be first in the queue.

I'd remove it, since it can confuse people into thinking that they can use a different timeout.

File: runtime/near-vm-runner/src/logic/logic.rs [hunk 7]
```diff
@@ -3435,6 +3435,80 @@ bls12381_p2_decompress_base + bls12381_p2_decompress_element * num_elements`
         Ok(new_promise_idx)
     }
```
Comment (jancionear) [inline runtime/near-vm-runner/src/logic/logic.rs:3435]: I wonder if we need the nearVM code if we're switching to wasmtime

File: runtime/runtime/src/ext.rs [hunk 8]
```diff
@@ -363,6 +364,52 @@ impl<'a> External for RuntimeExt<'a> {
...
+        {
+            return Err(HostError::YieldIdAlreadyExists.into());
+        }
+
+        // Check for duplicate in current transaction's pending receipts
+        // (yield_id maps to data_id, but we don't know the data_id yet —
+        // we need to check if any pending yield already used this yield_id.
+        // Since we store the mapping in trie above, this is handled by the trie check.
+        // But for yields created in the same function call, the trie hasn't been committed yet.
+        // We'll store a local mapping for that purpose.)
```
Comment (jancionear) [inline runtime/runtime/src/ext.rs]: This comment is confusing, the fact that the trie hasn't been commited yet doesn't matter, things written to the trie are immediately readable within the same function.

File: runtime/near-vm-runner/src/wasmtime_runner/logic.rs [hunk 9]
```diff
@@ -3774,6 +3774,107 @@ pub fn promise_yield_create(
...
+    yield_timeout_blocks: u64,
+) -> Result<u64> {
+    ctx.result_state.gas_counter.pay_base(base)?;
+    if ctx.context.is_view() {
+        return Err(HostError::ProhibitedInView {
+            method_name: "promise_yield_create_with_id".to_string(),
+        }
+        .into());
+    }
+    ctx.result_state.gas_counter.pay_base(yield_create_base)?;
```
Comment (jancionear) [inline runtime/near-vm-runner/src/wasmtime_runner/logic.rs]: Maybe the `yield_id` version should be a bit more expensive? It stores more data in the trie and does more trie operations. Probably doesn't matter much though, we could keep it the same.

File: runtime/near-vm-runner/src/wasmtime_runner/logic.rs [hunk 10]
```diff
@@ -3847,6 +3948,55 @@ pub fn promise_yield_resume(
     ctx.ext.submit_promise_resume_data(data_id, payload).map(u32::from)
 }
 
+/// Like [`promise_yield_resume`], but accepts the user-provided `yield_id` (from
+/// [`promise_yield_create_with_id`]) instead of the runtime-generated `data_id`.
+pub fn promise_yield_resume_with_id(
```
Comment (jancionear) [inline runtime/near-vm-runner/src/wasmtime_runner/logic.rs]: The `with_id` naming feels a bit confusing. It sounded better to me with three functions
```rust
// Create, generate id automatically
promise_yield_create
// Create, provide id manually
promise_yield_create_with_id
// resume some id
promise_yield_resume
```

But now we have `promise_yield_resume_with_id`, which is confusing because the normal `promise_yield_resume` already takes an id. Maybe it could be `with_yield_id`? A bit longer, but a bit clearer?

File: runtime/near-test-contracts/test-contract-rs/src/lib.rs [hunk 11]
```diff
@@ -1290,6 +1309,146 @@ pub unsafe fn call_yield_create_and_resume() {
...
+        yield_id.as_ptr() as u64,
+        yield_timeout_blocks,
+    );
+}
+
+/// Call promise_yield_resume_with_id with the user-provided yield_id.
+/// Input is a 32-byte yield_id followed by the payload.
+#[cfg(feature = "nightly")]
+#[unsafe(no_mangle)]
+pub unsafe fn call_yield_resume_with_id() {
```
Comment (jancionear) [inline runtime/near-test-contracts/test-contract-rs/src/lib.rs]: We have `call_promise` function in the test contract which you can use to do simple things like calling some host functions and creating a promise using a json argument. Could be easier than writing the functions manually, and it keeps the test contract smaller.

**Review by jancionear — COMMENTED** [review]
Overall looks good to me :+1:

I'll have a deeper look later once I'm done with BEEing

File: runtime/runtime/src/ext.rs [hunk 12]
```diff
@@ -363,6 +364,52 @@ impl<'a> External for RuntimeExt<'a> {
...
+        set_yield_id_mapping(&mut self.trie_update, &receiver_id, user_yield_id, input_data_id);
+
+        let receipt_index = self.receipt_manager.create_promise_yield_receipt_with_id(
+            input_data_id,
+            receiver_id.clone(),
+            yield_timeout_blocks,
+        );
+        let receipt_index = (receipt_index, input_data_id);
+
+        if ProtocolFeature::YieldResumeImprovements.enabled(self.current_protocol_version) {
```
Comment (jancionear) [inline runtime/runtime/src/ext.rs]: `ProtocolFeature::YieldResumeImprovements` is already live on mainnet, feel free to treat `enabled` as `true`. We should go and deprecate the feature at some point.

File: runtime/runtime/src/ext.rs [hunk 13]
```diff
@@ -401,6 +448,20 @@ impl<'a> External for RuntimeExt<'a> {
...
+        data: Vec<u8>,
+    ) -> Result<bool, VMLogicError> {
+        let Some(data_id) =
+            get_data_id_for_yield_id(self.trie_update, &self.account_id, user_yield_id)
+                .map_err(wrap_storage_error)?
+        else {
+            return Ok(false);
+        };
+        self.submit_promise_resume_data(data_id, data)
+    }
```
Comment (jancionear) [inline runtime/runtime/src/ext.rs:433]: AFAIU it is still possible to do `promise_yield_create_with_id` and then resume it using `promise_yield_resume` with the `data_id`. Sounds okay, but we should make it explicit in the documentation/tests.

File: runtime/near-test-contracts/test-contract-rs/src/lib.rs (hunk 4 above)
Comment (graphite) [reply runtime/near-test-contracts/test-contract-rs/src/lib.rs]: The original reasoning was that unless we have the timeout baked into the `promise_yield_create2` we will need to introduce yet another host function when we want to add it. I think it also implicitly assumed that `promise_yield_create2` will be strictly superior and `promise_yield_create` become legacy.

Thinking more about it I believe there will always be use cases for `promise_yield_create`, e.g. when developers do not want to care about `yield_id` and want the protocol to simply create something unique. We also moved away from `2` and named it explicitly, so we no longer have an argument that v2 allows custom IDs and timeouts. On the flip side once we introduce timeouts we may need them both for the original `promise_yield_create` and for the new one, making it 4 total host functions :laughing:.

All in all, I agree that adding a non-implemented feature with no specific use case that developers ask for is not the best idea, let me roll it back.

File: runtime/near-vm-runner/src/logic/logic.rs (hunk 7 above)
Comment (graphite) [reply runtime/near-vm-runner/src/logic/logic.rs:3435]: I asked the same on Zulip and we agreed that we will remove all the nearVM tests, but only after wasmtime switch is fully complete and up to then it is beneficial to keep the tests for nearVM.

File: runtime/runtime/src/function_call.rs (hunk 6 above)
Comment (graphite) [reply runtime/runtime/src/function_call.rs]: As discussed in the other comment, rolling the whole custom timeout back.

File: runtime/near-test-contracts/test-contract-rs/src/lib.rs (hunk 5 above)
Comment (graphite) [reply runtime/near-test-contracts/test-contract-rs/src/lib.rs]: Good point, we should leave it to the contract, let me patch.

Comment (graphite) [reply runtime/near-test-contracts/test-contract-rs/src/lib.rs]: One thing I ran into here is that our canonical way to signal an error is return 0. At the same time 0 is a valid promise_idx, so instead I am returning `u64::MAX`. This needs to be clearly documented and covered by the SDK.

File: core/primitives/src/trie_key.rs (hunk 3 above)
Comment (graphite) [reply core/primitives/src/trie_key.rs]: When I thought about it I decided that it should be handled on the SDK side, but I overlooked that we may confuse these two in nearcore code base as well. Adding.

File: runtime/near-vm-runner/src/wasmtime_runner/logic.rs (hunk 9 above)
Comment (graphite) [reply runtime/near-vm-runner/src/wasmtime_runner/logic.rs]: It does 2 more trie writes, so based on that I estimated it at 0.29 Tgas vs. .153 Tgas for the original `promise_yield_create`. I also ran params-estimator, but that came back with ~0.05 Tgas, so I keep it at .29 Tgas. cc @mitinarseny in case you think this is too expensive to be useful.

I additionally considered pricing the new resume differently as it performs an additional read, plus needs to handle the mapping cleanup. However, the mapping cleanup needs to happen in the "create with yield id, resume by data id" as well, and the estimated diff is within 5%, which is likely within the uncertainty margin anyway.

**Comment by graphite** [comment]
@jancionear addressed your comments, great feedback! I left several conversations open in case you would like to follow up

File: chain/jsonrpc/openapi/src/main.rs [hunk 14]
```diff
@@ -643,7 +643,7 @@ fn whole_spec(all_schemas: SchemasMap, all_paths: PathsMap) -> OpenApi {
         openapi: "3.0.0".to_string(),
         info: okapi::openapi3::Info {
             title: "NEAR Protocol JSON RPC API".to_string(),
-            version: "1.2.6".to_string(),
+            version: "1.2.7".to_string(),
```
Comment (jancionear) [inline chain/jsonrpc/openapi/src/main.rs]: I'm not sure what are the rules for version bumps, but it makes sense to me :+1:

File: runtime/near-test-contracts/build.rs [hunk 15]
```diff
@@ -23,7 +23,7 @@ fn main() {
 fn try_main() -> Result<(), Error> {
     let mut test_contract_features = vec![];
 
-    let is_nightly = std::env::var_os("CARGO_FEATURE_nightly").is_some();
+    let is_nightly = std::env::var_os("CARGO_FEATURE_NIGHTLY").is_some();
```
Comment (jancionear) [inline runtime/near-test-contracts/build.rs:26]: ouch, nice that this is fixed

File: runtime/near-vm-runner/src/profile.rs [hunk 16]
```diff
@@ -332,7 +332,7 @@ mod test {
             storage_remove_ret_value_byte -> 35 [0% host]
             storage_has_key_base -> 36 [0% host]
             storage_has_key_byte -> 37 [0% host]
-            storage_iter_create_prefix_base -> 38 [1% host]
+            storage_iter_create_prefix_base -> 38 [0% host]
```
Comment (jancionear) [inline runtime/near-vm-runner/src/profile.rs]: Why those changed?

File: core/parameters/res/runtime_configs/154.yaml [hunk 17]
```diff
@@ -0,0 +1,6 @@
+yield_with_id_host_fns: { old: false, new: true }
+# yield_create_with_id charges this base instead of wasm_yield_create_base.
+# Accounts for the two extra trie writes (YieldIdToDataId / DataIdToYieldId
+# mapping) compared to plain yield_create. Value derived from wasm_yield_create_base
+# (153B) plus the cost of two wasm_storage_write_base operations (~68B each).
+wasm_yield_create_with_id_base: { old: 300_000_000_000_000, new: 290_000_000_000 }
```
Comment (jancionear) [inline core/parameters/res/runtime_configs/154.yaml]: nit: wasm_storage_write_base is 64B

File: core/primitives/src/profile_data_v3.rs [hunk 18]
```diff
@@ -373,6 +373,7 @@ mod test {
             storage_large_read_overhead_byte -> 84 [2% host]
             p256_verify_base -> 85 [2% host]
             p256_verify_byte -> 86 [2% host]
+            yield_create_with_id_base -> 87 [2% host]
             ------ Actions --------
             create_account -> 1000
             delete_account -> 1001
```
Comment (jancionear) [inline core/primitives/src/profile_data_v3.rs:379]: I'm not sure how the estimator works exactly, but looks good to me :+1:

File: runtime/near-vm-runner/src/wasmtime_runner/logic.rs [hunk 19]
```diff
@@ -3772,6 +3772,107 @@ pub fn promise_yield_create(
...
+        &ctx.fees_config,
+        ActionCosts::function_call_byte,
+        num_bytes,
+        true,
+    )?;
+    ctx.ext.append_action_function_call_weight(
+        new_receipt_idx,
+        method_name,
+        arguments,
+        Balance::ZERO,
```
Comment (jancionear) [inline runtime/near-vm-runner/src/wasmtime_runner/logic.rs]: It's interesting that we don't allow to attach any balance to yielded receipts. Maybe it makes sense, it's the same account.

File: runtime/near-vm-runner/src/logic/tests/yield_resume.rs [hunk 20]
```diff
@@ -0,0 +1,573 @@
+use crate::logic::HostError;
+use crate::logic::mocks::mock_external::MockAction;
+use crate::logic::tests::vm_logic_builder::VMLogicBuilder;
+use near_primitives_core::hash::CryptoHash;
+use near_primitives_core::version::{PROTOCOL_VERSION, ProtocolFeature};
+
+#[test]
+fn test_promise_yield_create() {
+    let mut logic_builder = VMLogicBuilder::free();
```
Comment (jancionear) [inline runtime/near-vm-runner/src/logic/tests/yield_resume.rs]: I think these tests are testing the `MockExternal`, not the real implementation, not sure if that was the intention.

All important test should use `RuntimeNode` (as in `integration-tests/src/tests/runtime/test_yield_resume.rs`) or a whole TestLoop network (e.g. `test-loop-tests/src/tests/yield_resume.rs`)

File: core/primitives/src/trie_key.rs [hunk 21]
```diff
@@ -76,10 +79,14 @@ pub mod col {
     // GasKeyNonce rows live on disk under `ACCESS_KEY` (extending the access-key
     // trie key with a `NonceIndex` suffix), so no `col` constant exists for byte
     // 21. Do not introduce one without coordinating with the `TrieKey` repr.
+    /// Mapping from user-provided yield ID to runtime data ID.
+    pub const YIELD_ID_TO_DATA_ID: u8 = 22;
+    /// Reverse mapping from runtime data ID to user-provided yield ID.
+    pub const DATA_ID_TO_YIELD_ID: u8 = 23;
```
Comment (jancionear) [inline core/primitives/src/trie_key.rs:85]: These columns are not handled in `shard_split_handle_key_value`, this will probably cause a panic on resharding.

It would be good to add a resharding test that yields before the resharding boundary and resumes after. Something similar to https://github.com/near/nearcore/pull/15754 or what was done in https://github.com/near/nearcore/pull/15085. It might be a bit tricky, we don't really have good test infrastructure for that :/

**Review by jancionear — APPROVED** [review]
LGTM :+1:

File: core/primitives/src/profile_data_v3.rs (hunk 18 above)
Comment (graphite) [reply core/primitives/src/profile_data_v3.rs:379]: Not sure either, I ran it and got _some_ output.

File: runtime/near-vm-runner/src/logic/tests/yield_resume.rs (hunk 20 above)
Comment (graphite) [reply runtime/near-vm-runner/src/logic/tests/yield_resume.rs]: IIUC the existing tests under `integration-tests` will test these test cases as well as long as they are running with the correct runtime. I'll drop the tests that are testing `MockExternal` and keep only the ones that do not call to it, expecting `near-vm-runner` tests to become obsolete soon.

File: runtime/near-vm-runner/src/wasmtime_runner/logic.rs (hunk 19 above)
Comment (mitinarseny) [reply runtime/near-vm-runner/src/wasmtime_runner/logic.rs]: @jancionear thank you for spotting it! @graphite Can we please allow that? We use this trick [here](https://github.com/near/intents/blob/13485834520f0ca7bb3013535d39b00bd7dc6b14/contracts/global-deployer/src/contract.rs#L51-L62)

Comment (jancionear) [reply runtime/near-vm-runner/src/wasmtime_runner/logic.rs]: Could you explain a bit more what the trick does and how attaching balance would help there?

Comment (mitinarseny) [reply runtime/near-vm-runner/src/wasmtime_runner/logic.rs]: The use-case is following: sometimes you need to have a public contract method where you want to accept some attached deposit, but only do some actions with it later (e.g. when yielded promise gets resumed). At the same time, you don't want this attached deposit to be added to your contract's balance until this promise is resolved, because:
1. `env::account_balance()` might take part in calculations for other "concurrent" method calls on your contract from other predecessors
2. You don't want to store any data indicating "this amount is resolved for this previously yielded callback promise" in your contract's storage - we can allow yield API to temporarily store it in the yielded receipt itself

Since function call in the yielded promise is the first action that will be executed, you can't prepend it with another "transfer to myself" action - the current yield API doesn't allow for that. If we add `attached_deposit` field - this would become possible and enable new use-cases.

Comment (graphite) [reply runtime/near-vm-runner/src/wasmtime_runner/logic.rs]: Added to the new variant, adjusting the existing is probably not possible at all.

**Comment by graphite** [comment]
@jancionear addressed your and copilot's feedback, plus the request to allow balance attached. This took 9 commits, but all are pretty small and easy to read, so it may make sense to quickly look at them. Probably the one with the resharding test and nightly features deserves some attention.

File: runtime/near-vm-runner/src/logic/logic.rs [hunk 22]
```diff
@@ -3433,6 +3433,87 @@ bls12381_p2_decompress_base + bls12381_p2_decompress_element * num_elements`
...
+        if self.context.is_view() {
+            return Err(HostError::ProhibitedInView {
+                method_name: "promise_yield_create_with_id".to_string(),
+            }
+            .into());
+        }
+        self.result_state.gas_counter.pay_base(yield_create_with_id_base)?;
+        let amount = Balance::from_yoctonear(
+            self.memory.get_u128(&mut self.result_state.gas_counter, amount_ptr)?,
+        );
```
Comment (jancionear) [inline runtime/near-vm-runner/src/logic/logic.rs:3465]: Isn't that missing `self.result_state.deduct_balance(amount)?;` ? Looks like it doesn't deduct the attached tokens from account balance and mints them out of thin air :eyes:

**Review by jancionear — CHANGES_REQUESTED** [review]
I think there's an issue with attaching balance.

Btw `git merge master` is nicer than `git rebase master` because it's easier for the reviewer to compare what changed between versions.

File: runtime/near-vm-runner/src/logic/logic.rs (hunk 22 above)
Comment (jancionear) [reply runtime/near-vm-runner/src/logic/logic.rs:3465]: It would be good to add a test which sends some amount and verifies that it was deducted from the account.

Comment (jancionear) [reply runtime/near-vm-runner/src/logic/logic.rs:3465]: We could also support 1 yoctonear

Comment (graphite) [reply runtime/near-vm-runner/src/logic/logic.rs:3465]: @jancionear Fixed and handling 1 yocto as well plus added a bunch of tests, please check.

File: runtime/near-vm-runner/src/logic/tests/yield_resume.rs [hunk 23]
```diff
@@ -289,3 +290,183 @@ fn test_promise_yield_resume_with_yield_id_view_prohibited() {
...
+    )
+}
+
+#[test]
+fn test_promise_yield_create_with_id_deducts_balance() {
+    if !ProtocolFeature::YieldWithId.enabled(PROTOCOL_VERSION) {
+        return;
+    }
+
+    let mut logic_builder = VMLogicBuilder::free();
```
Comment (jancionear) [inline runtime/near-vm-runner/src/logic/tests/yield_resume.rs:246]: These tests use `MockExternal` again :c

It would be good to have an integration test (`integration-tests/src/tests/runtime/test_yield_resume.rs`) which attaches some amount, makes sure the balance is deduced, then added back, etc.
Previously it was impossible to have a `PromiseYield` receipt with balance, now we have them and it could trigger some unexpected paths in the runtime code. From a brief look it looks like it should be fine, but it would be good to confirm with a test.

**Review by jancionear — APPROVED** [review]
LGTM

code looks good to me. Could use a few more tests for deposit attaching.

File: runtime/near-vm-runner/src/logic/tests/yield_resume.rs (hunk 23 above)
Comment (graphite) [reply runtime/near-vm-runner/src/logic/tests/yield_resume.rs:246]: Removed all the tests that test `MockExternal`, the remaining ones test VM logic AFAIU.

I added the same as integration tests, but it is a bit trickier than I expected:
 - 30% gas rebates make it harder to check the exact balance
 - Testing 1 yocto exception requires zero balance, which is only possible with ZBAs, so had to switch it to a minimal global contract

File: integration-tests/src/tests/runtime/test_yield_resume.rs [hunk 24]
```diff
@@ -443,3 +469,322 @@ fn create_then_resume_with_yield_id_fails() {
...
+    let gas_burnt_for_function_call = res.receipts_outcome[0]
+        .outcome
+        .gas_burnt
+        .checked_sub(fee_helper.function_call_exec_gas(num_bytes))
+        .unwrap();
+    fee_helper.gas_burnt_to_reward(gas_burnt_for_function_call)
+}
+
+#[test]
+#[cfg(feature = "nightly")]
```
Comment (jancionear) [inline integration-tests/src/tests/runtime/test_yield_resume.rs]: nit: something like this would make things easier during stabilization:

```rust
#[test]
fn create_with_id_deducts_attached_amount() {
    if !ProtocolFeature::YieldWithId.enabled(PROTOCOL_VERSION) {
        return;
    }
```

With `#[cfg(feature = "nightly")]` we will have to remember to re-enable all of the tests on stable later.

File: integration-tests/src/tests/runtime/test_yield_resume.rs [hunk 25]
```diff
@@ -443,3 +469,322 @@ fn create_then_resume_with_yield_id_fails() {
...
+    );
+
+    let reward = contract_function_call_reward(&fee_helper(&node), &res, "call_promise", args_len);
+    let view = node.user().view_account(&zba_id).unwrap();
+    let expected = balance_before.checked_add(reward).unwrap();
+    assert_eq!(
+        view.amount, expected,
+        "zba balance must equal pre-call balance plus the function-call gas reward",
+    );
+}
```
Comment (jancionear) [inline integration-tests/src/tests/runtime/test_yield_resume.rs:796]: Would be nice to also have a test where the receipt is resumed and the balance should reappear on the account.

**Review by jancionear — APPROVED** [review]

File: integration-tests/src/tests/runtime/test_yield_resume.rs (hunk 25 above)
Comment (graphite) [reply integration-tests/src/tests/runtime/test_yield_resume.rs:796]: I tried to achieve this, but it becomes extremely convoluted very quickly, there are several receipts produced and it is very hard if at all possible to accurately calculate the gas rebate, which is needed for tracking the expected balance.

File: integration-tests/src/tests/runtime/test_yield_resume.rs (hunk 24 above)
Comment (graphite) [reply integration-tests/src/tests/runtime/test_yield_resume.rs]: Done
